File: ts/solutions/iterables/iterables-6.ts
// O(n)
function compressString(s: string): string {
  if (s === '') {
    return ''
  }

  const acc: string[] = []
  let startIndex = 0
  let currentIndex = 0
  let currentChar = s[0]

  for (const c of s) {
    if (c !== currentChar) {
      acc.push(currentChar, '' + (currentIndex - startIndex))
      startIndex = currentIndex
      currentChar = c
    }
    currentIndex++
  }

  acc.push(currentChar, '' + (currentIndex - startIndex))

  return acc.length > s.length ? s : acc.join('')
}

console.log(compressString('aabcccccaaa'))

File: python/iterables/iterables_6.py
# O(n)
def compress_string(s):
    # handle edge case ahead of assignment to s[0]
    if s == '':
        return ''

    # accumulator list to store compressed string
    acc = []

    start_index = 0
    current_index = 0
    current_char = s[0]

    # O(n)
    for c in s:
        if c != current_char:
            # O(1)
            acc.extend([current_char, str(current_index - start_index)])
            start_index = current_index
            current_char = c
        current_index += 1

    acc.extend([current_char, str(current_index - start_index)])

    # O(n)
    return ''.join(acc) if len(acc) < len(s) else s
